File: MyAppOne/MyAppOne/VC/Detail/MAODetailViewController.m
//
//  MAODetailViewController.m
//  MyAppOne
//
//  Created by Roman Guarino on 17/01/2019.
//  Copyright © 2019 iOS-accelerator. All rights reserved.
//

#import "MAODetailViewController.h"
#import "../../Model/MAOListViewControllerModel.h"

@interface MAODetailViewController ()
@property (weak, nonatomic) IBOutlet UILabel *lblTrackName;
@property (weak, nonatomic) IBOutlet UILabel *lblAlbumName;
@property (weak, nonatomic) IBOutlet UILabel *lblAutorName;
@property (weak, nonatomic) IBOutlet UIImageView *imgAlbum;
@property (weak, nonatomic) IBOutlet UILabel *lblPriceTrack;
@property (weak, nonatomic) IBOutlet UIView *contentView;
@property (nonatomic, strong) MAOListViewControllerModel *model;
@end

@implementation MAODetailViewController

- (instancetype) initWithModel:(MAOListViewControllerModel *) model {
    self = [super init];
    if(self){
        _model = model;
    }
    return self;
}

- (void)viewDidLoad {
    self.view.backgroundColor = [UIColor colorWithWhite:0.0f alpha:0.80];
    _contentView.layer.cornerRadius = 10;
    
    _lblTrackName.text = _model.trackName;
    _lblAlbumName.text = _model.collectionName;
    _lblAutorName.text = _model.artistName;
    _lblPriceTrack.text = [NSString stringWithFormat:@"$ %@", _model.trackPrice];
    
    dispatch_queue_t queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0ul);
    dispatch_async(queue, ^{
        NSData *data = [NSData dataWithContentsOfURL:[NSURL URLWithString:self.model.artworkUrl100]];
        dispatch_async(dispatch_get_main_queue(), ^{
            self.imgAlbum.image = [UIImage imageWithData:data];
        });
    });
    
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    if(touches.anyObject.view != self.contentView) {
        [self dismissViewControllerAnimated:YES completion:nil];
    }
}

- (IBAction)btnClose:(UIButton *)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (IBAction)btnBuy:(UIButton *)sender {
    [[UIApplication sharedApplication]
       openURL:[NSURL URLWithString:_model.previewUrl] options:@{} completionHandler:nil];
}

@end

File: MyAppOneSwift/MyAppOneSwift/VC/Detail/MAODetailViewController.swift
//
//  MAODetailViewController.swift
//  MyAppOneSwift
//
//  Created by Roman Guarino on 28/01/2019.
//  Copyright © 2019 Román Guarino. All rights reserved.
//

import UIKit

class MAODetailViewController: UIViewController {

    @IBOutlet weak var lblTrackName: UILabel!
    @IBOutlet weak var lblAlbumName: UILabel!
    @IBOutlet weak var lblAutorName: UILabel!
    @IBOutlet weak var lblTrackPrice: UILabel!
    @IBOutlet weak var imgAlbum: UIImageView!
    @IBOutlet weak var contentView: UIView!
    private var model: MAOListViewControllerModel!
    
    init(model: MAOListViewControllerModel){
        self.model = model
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        contentView.layer.cornerRadius = 10
        self.view.backgroundColor = UIColor.init(white: 0.0, alpha: 0.8)
        
        lblTrackName.text = model.trackName
        lblAlbumName.text = model.collectionName
        lblAutorName.text = model.artistName
        lblTrackPrice.text = model.trackPrice?.description
        DispatchQueue.global().async { [weak self] in
            if let data = try? Data(contentsOf: URL.init(string: self!.model.artworkUrl100!)!) {
                if let image = UIImage(data: data) {
                    DispatchQueue.main.async {
                        self?.imgAlbum.image = image
                    }
                }
            }
        }
    }
    
    @IBAction func btnClose(_ sender: UIButton) {
        self.dismiss(animated: true) {}
    }
    
    @IBAction func btnBuy(_ sender: UIButton) {
        UIApplication.shared.open(URL(string: (model.previewUrl)!)!, options: [:]) { (Bool) in }
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        if(touches.first?.view != self.contentView) {
            self.dismiss(animated: true) {}
        }
    }
}
